File: java-lambda/src/main/java/jalexander/ninja/Merge.java
package jalexander.ninja;

import com.amazonaws.services.lambda.runtime.Context;

import java.util.Arrays;

public class Merge {
    public static class RequestClass {
        public int inputInt;
        public String inputString;
    }

    public static class ResponseClass {
        public int outputInt;
        public double time1;
        public double time2;
        public long longTime1;
        public long longTime2;
        public double outputDouble;
        public String outputString;
    }

    private static class MergeSorter extends Thread {
        public int[] arr;
        public int depth;
        public static int started;
        final static int THREAD_DEPTH = 3;

        public MergeSorter(int[] a, int d) {
            arr = a;
            depth = d;
        }

        public void run() {
            if (arr.length <= 1) {
                return;
            }

            if (arr.length == 2) {
                if (arr[1] < arr[0]) {
                    int t = arr[1];
                    arr[1] = arr[0];
                    arr[0] = t;
                }
                return;
            }

            int split = arr.length / 2;
            MergeSorter left = new MergeSorter(Arrays.copyOfRange(arr, 0, split), depth + 1);
            if (depth <= THREAD_DEPTH) {
                left.start();
                started++;
            } else {
                left.run();
            }
            MergeSorter right = new MergeSorter(Arrays.copyOfRange(arr, split, arr.length), depth + 1);
            right.run();

            if (depth <= THREAD_DEPTH) {
                try {
                    left.join();
                } catch (InterruptedException e) {
                    System.err.println("Interrupted in MergeSorter");
                    System.exit(1);
                }
            }

            int lPos = 0;
            int rPos = 0;
            int cPos = 0;

            while (lPos < left.arr.length || rPos < right.arr.length) {
                if (rPos >= right.arr.length) {
                    arr[cPos] = left.arr[lPos];
                    lPos++;
                } else if (lPos >= left.arr.length) {
                    arr[cPos] = right.arr[rPos];
                    rPos++;
                } else if (left.arr[lPos] < right.arr[rPos]) {
                    arr[cPos] = left.arr[lPos];
                    lPos++;
                } else {
                    arr[cPos] = right.arr[rPos];
                    rPos++;
                }
                cPos++;
            }
        }
    }

    public static void mergeSort(int[] arr) {
        MergeSorter m = new MergeSorter(arr, 0);
        m.start();
        try {
            m.join();
        } catch (InterruptedException e) {
            System.err.println("Interrupted in mergeSort");
            System.exit(1);
        }

        /*
        for (int i = 0; i < arr.length; i++) {
            System.out.print(arr[i] + " ");
        }
        System.out.println("Started: " + MergeSorter.started);
        */
    }

    public static ResponseClass handler(RequestClass request, Context context){
        ResponseClass response = new ResponseClass();
        response.outputString = "perfomed merge sort with " + request.inputInt + " elements";

        long start2 = System.currentTimeMillis();
        int[] arr = new int[request.inputInt];
        for(int i = 0; i < arr.length; i++){
            arr[i] = (int)(Math.random() * Integer.MAX_VALUE);
        }

        long start = System.currentTimeMillis();
        mergeSort(arr);
        long end = System.currentTimeMillis();

        response.longTime1 = end - start;
        response.longTime2 = end - start2;
        return response;
    }
}

File: go-lambda/merge.go
package main

import (
	"fmt"
	"github.com/aws/aws-lambda-go/lambda"
	"math/rand"
	"os"
	"strconv"
	"time"
)

type RequestClass struct {
	InputInt    int    `json:"inputInt"`
	InputString string `json:"inputString"`
}

type ResponseClass struct {
	OutputInt    int     `json:"outputInt"`
	Time1        float64 `json:"time1"`
	Time2        float64 `json:"time2"`
	LongTime1    int64   `json:"longTime1"`
	LongTime2    int64   `json:"longTime2"`
	OutputDouble float64 `json:"outputDouble"`
	OutputString string  `json:"outputString"`
}

func mergeSortSync(arr []int, depth int) []int {
	const THREAD_DEPTH = 3

	if len(arr) <= 1 {
		return arr
	}

	if len(arr) == 2 {
		if arr[1] < arr[0] {
			t := arr[1]
			arr[1] = arr[0]
			arr[0] = t
			return arr
		}
	}

	split := len(arr) / 2

	left := make([]int, split)
	right := make([]int, len(arr)-split)

	copy(left, arr[:split])
	copy(right, arr[split:])

	ch := make(chan []int)
	var leftRes []int

	if depth <= THREAD_DEPTH {
		go mergeSortAsync(left, depth+1, ch)
	} else {
		leftRes = mergeSortSync(left, depth+1)
	}

	rightRes := mergeSortSync(right, depth+1)

	if depth <= THREAD_DEPTH {
		leftRes = <-ch
	}

	lPos := 0
	rPos := 0
	cPos := 0

	for lPos < len(leftRes) || rPos < len(rightRes) {
		if rPos >= len(rightRes) {
			arr[cPos] = leftRes[lPos]
			lPos++
		} else if lPos >= len(leftRes) {
			arr[cPos] = rightRes[rPos]
			rPos++
		} else if leftRes[lPos] < rightRes[rPos] {
			arr[cPos] = leftRes[lPos]
			lPos++
		} else {
			arr[cPos] = rightRes[rPos]
			rPos++
		}
		cPos++
	}

	return arr
}

func mergeSortAsync(arr []int, depth int, c chan []int) {
	c <- mergeSortSync(arr, depth)
}

func Handler(input RequestClass) (ResponseClass, error) {
	response := ResponseClass{}

	rand.Seed(time.Now().UTC().UnixNano())
	start2 := time.Now()

	arr := make([]int, input.InputInt)
	for i := 0; i < input.InputInt; i++ {
		//arr[i] = rand.Intn(2147483647)
		arr[i] = rand.Intn(100)
	}

	start := time.Now()
	mergeSortSync(arr, 0)
	end := time.Now()

	response.LongTime1 = int64(end.Sub(start) / time.Millisecond)
	response.LongTime2 = int64(end.Sub(start2) / time.Millisecond)

	response.OutputString = fmt.Sprintf("(go) perfomed merge sort with %d elements", input.InputInt)

	return response, nil
}

func main() {
	if len(os.Args) == 3 {
		var req RequestClass
		input, err := strconv.Atoi(os.Args[1])
		req.InputInt = input
		req.InputString = os.Args[2]
		output, err := Handler(req)
		err = err
		fmt.Println(output)
	} else {
		lambda.Start(Handler)
	}
}
